### PR TITLE
Proofs for Detype_R.thy, Retype_R.thy, Untyped_R.thy and more

### PR DESCRIPTION
**Pull request by michaelmcinerney**

### CODE REVIEW

**Review by lsf37 — APPROVED** [review]
Looks good! No objections from me.

File: proof/refine/RISCV64/InterruptAcc_R.thy [hunk 1]
```diff
@@ -94,37 +89,290 @@ lemma work_units_and_irq_state_state_relationI [intro!]:
...
+   apply (fastforce dest: valid_objs_valid_sched_context_size
+                    simp: sc_at_pred_n_def obj_at_def is_sc_obj_def)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma modifyWorkUnits_valid_objs'[wp]:
+  "modifyWorkUnits f \<lbrace>valid_objs'\<rbrace>"
+  apply (clarsimp simp: modifyWorkUnits_def)
+  apply wpsimp
+  done
```
Comment (Xaphiosis) [inline proof/refine/RISCV64/InterruptAcc_R.thy:178]: nitpick: nicer style (for future consideration) might be
```
unfolding modifyWorkUnits_def
by wpsimp
```

File: proof/refine/RISCV64/VSpace_R.thy [hunk 2]
```diff
@@ -12,6 +12,13 @@ theory VSpace_R
 imports TcbAcc_R
 begin
 
+context Arch begin global_naming RISCV64 (*FIXME: arch_split*)
+
+(*FIXME: move to ainvs*)
```
Comment (Xaphiosis) [inline proof/refine/RISCV64/VSpace_R.thy:17]: FIXME?

File: proof/refine/RISCV64/VSpace_R.thy [hunk 3]
```diff
@@ -325,10 +332,39 @@ lemma deleteASIDPool_corres:
...
+sublocale Arch < setMRs: typ_at_all_props' "setMRs thread buffer messageData"
+  by typ_at_props'
+
+sublocale Arch < performPageTableInvocation: typ_at_all_props' "performPageTableInvocation iv"
+  by typ_at_props'
+
+sublocale Arch < performASIDPoolInvocation: typ_at_all_props' "performASIDPoolInvocation iv"
+  by typ_at_props'
+
+sublocale Arch < performPageInvocation: typ_at_all_props' "performPageInvocation iv"
```
Comment (Xaphiosis) [inline proof/refine/RISCV64/VSpace_R.thy:364]: I'm curious, what does this do?

File: proof/refine/RISCV64/Retype_R.thy [hunk 4]
```diff
@@ -3685,20 +3786,21 @@ lemma createNewCaps_state_refs_of':
...
+                           pteBits_def makeObject_sc makeObject_reply pte_bits_def pt_bits_def
+                           table_size_def ptTranslationBits_def word_size_bits_def
+               | intro conjI impI )+
 
 lemma createObjects_iflive':
   "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> \<not> live' val
         \<and> n \<noteq> 0
         \<and> range_cover ptr sz (objBitsKO val + gbits) n
-        \<and> pspace_aligned' s \<and> pspace_distinct' s
+        \<and> pspace_aligned' s \<and> pspace_distinct' s \<and> pspace_bounded' s
```
Comment (Xaphiosis) [inline proof/refine/RISCV64/Retype_R.thy]: for the ignorant: what does pspace_bounded represent?

File: proof/refine/RISCV64/VSpace_R.thy (hunk 3 above)
Comment (michaelmcinerney) [reply proof/refine/RISCV64/VSpace_R.thy:364]: The `typ_at_all_props'` locale will take the `typ_at'` and `sc_at'_n` lemmas we have and will generate all the `tcb_at'`, `ep_at'` type lemmas

File: proof/refine/RISCV64/Retype_R.thy (hunk 4 above)
Comment (michaelmcinerney) [reply proof/refine/RISCV64/Retype_R.thy]: It says that the size of everything in the heap is bounded by `word_bits`. In full,

`pspace_bounded' s \<equiv>
  \<forall>x \<in> dom (ksPSpace s). objBitsKO (the (ksPSpace s x)) < word_bits`

File: proof/refine/RISCV64/VSpace_R.thy (hunk 3 above)
Comment (Xaphiosis) [reply proof/refine/RISCV64/VSpace_R.thy:364]: Cool, that's shiny new tech I'm not familiar with.

**Comment by Xaphiosis** [comment]
Nice work. There are a lot of changes overall, and a bunch of things that I don't get because I'm not familiar with the MCS proof base, but no obvious "this proof sucks" or style regrets jumped out at me.

**Review by Xaphiosis — APPROVED** [review]

**Pull request merged**